### PR TITLE
WIP: feat(compiling): Include compile linked files in compilation

### LINKED ISSUE #128: nuget contentfiles mutations causes issues
With nuget packages, you can include compilable source files: https://blog.nuget.org/20160126/nuget-contentFiles-demystified.html

These files end up compiled into the `obj\Debug\netcoreapp2.1\NuGet\F3F32FCC102FAF927C28583AC554AF5CC6C67A63\MyLibrary\0.1.2` folder of the project that consumes them. They are effectively linked-in source files.

When Stryker mutates these files, it fails to mutate them back again (presumably because nuget automatically resets them?)

```
[12:26:48 ERR] Unable to rollback mutation for node EmptyQuery with diagnosticmessage The name 'EmptyQuery' does not exist in the current context
[12:26:48 ERR] An error occurred during the mutationtest run
System.ArgumentException: Node to track is not a descendant of the root.
   at Microsoft.CodeAnalysis.SyntaxNodeExtensions.TrackNodes[TRoot](TRoot root, IEnumerable`1 nodes) in /_/src/Compilers/Core/Portable/Syntax/SyntaxNodeExtensions_Tracking.cs:line 43
   at Stryker.Core.Compiling.RollbackProcess.RemoveMutantIfStatements(SyntaxTree originalTree, ICollection`1 diagnosticInfo)
   at Stryker.Core.Compiling.RollbackProcess.Start(CSharpCompilation compiler, ImmutableArray`1 diagnostics)
   at Stryker.Core.Compiling.CompilingProcess.Compile(IEnumerable`1 syntaxTrees, MemoryStream ms)
   at Stryker.Core.MutationTest.MutationTestProcess.Mutate()
   at Stryker.Core.StrykerRunner.RunMutationTest(StrykerOptions options)
```

In the above message, `EmptyQuery` refers to a method defined in a base class inside the imported source contentFile from the nuget package.

It would be great if either compiled contentFiles from `obj` were ignored from mutation, and/or the ability to ignore files was added. The error looks similar to the error in #125

### LINKED ISSUE #523: Feature request: Support for linked files
**Is your feature request related to a problem? Please describe.**
My project uses `<Compile Include="..\MyAttributes.cs" Link="MyAttributes.cs" />` to include a file outside the project folder in the compilation.

When running Stryker 0.11.0 I get an error, that `MyAttributes.cs` cannot be found.

```
PS C:\Users\jonas\source\repos\StrykerCompileLink\StrykerTests> dotnet stryker --log-file

   _____ _              _               _   _ ______ _______
  / ____| |            | |             | \ | |  ____|__   __|
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|
                   __/ |
                  |___/


Version 0.11.0 (beta)

[20:44:15 INF] Using C:\Users\jonas\source\repos\StrykerCompileLink\StrykerTests\StrykerTests.csproj as project file
[20:44:19 INF] Started initial build using dotnet build
[20:44:21 INF] Initial build successful
[20:44:21 INF] Using testrunner DotnetTest
[20:44:21 INF] Initial testrun started
[20:44:24 INF] Total number of tests found in initial test run: 1
[20:44:24 INF] Initial testrun successful in 2270 ms
[20:44:24 INF] Using 33405 ms as testrun timeout
[20:44:26 WRN] Stryker.NET encountered an compile error in C:\Users\jonas\source\repos\StrykerCompileLink\StrykerCompileLink\Math.cs with message: The type or namespace name 'MyAttributes' could not be found (are you missing a using directive or an assembly reference?) (Source code: MyAttributes)
[20:44:26 WRN] Safe Mode! Stryker will try to continue by rolling back all mutations in method. This should not happen, please report this as an issue on github with the previous error message.
[20:44:26 FTL] Stryker.NET could not compile the project after mutation. This is probably an error for Stryker.NET and not your project. Please report this issue on github with the previous error message.
[20:44:26 ERR] An error occurred during the mutation test run
System.ApplicationException: Internal error due to compile error.
   at Stryker.Core.Compiling.RollbackProcess.RemoveMutantIfStatements(SyntaxTree originalTree, ICollection`1 diagnosticInfo, Boolean devMode) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\RollbackProcess.cs:line 188
   at Stryker.Core.Compiling.RollbackProcess.Start(CSharpCompilation compiler, ImmutableArray`1 diagnostics, Boolean devMode) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\RollbackProcess.cs:line 61
   at Stryker.Core.Compiling.CompilingProcess.TryCompilation(MemoryStream ms, CSharpCompilation compilation, EmitResult previousEmitResult, Boolean devMode, Int32 retryCount) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\CompilingProcess.cs:line 108
   at Stryker.Core.Compiling.CompilingProcess.Compile(IEnumerable`1 syntaxTrees, MemoryStream ms, Boolean devMode) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\CompilingProcess.cs:line 76
   at Stryker.Core.MutationTest.MutationTestProcess.Mutate(StrykerOptions options) in D:\a\1\s\src\Stryker.Core\Stryker.Core\MutationTest\MutationTestProcess.cs:line 91
   at Stryker.Core.StrykerRunner.RunMutationTest(StrykerOptions options) in D:\a\1\s\src\Stryker.Core\Stryker.Core\StrykerRunner.cs:line 77
[20:44:26 INF] Time Elapsed 00:00:10.6122125
```

**Describe the solution you'd like**
The project should compile.

**Describe alternatives you've considered**

**Additional context**
Full log: The compilation crashes before producing a log
Full example: 
[StrykerCompileLink.zip](https://github.com/stryker-mutator/stryker-net/files/3199368/StrykerCompileLink.zip)

### PR DESCRIPTION
**Pull request by Mobrockers**
Fix #523 
Fix #128

### CODE REVIEW

File: src/Stryker.Core/Stryker.Core/ProjectComponents/FolderComposite.cs [hunk 1]
```diff
@@ -27,7 +27,21 @@ public override IEnumerable<FileLeaf> GetAllFiles()
 
         public override void Add(ProjectComponent component)
         {
-            Children.Add(component);
+            if (component is FolderComposite folder && Children.Any(c => c.FullPath == folder.FullPath))
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/ProjectComponents/FolderComposite.cs]: Cheers to this method! 🎉

File: src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs [hunk 2]
```diff
@@ -96,15 +111,17 @@ private FolderComposite FindInputFiles(string path, List<string> filesToExclude,
...
-            }
-            foreach (var file in _fileSystem.Directory.GetFiles(folderComposite.FullPath, "*.cs", SearchOption.TopDirectoryOnly))
+
+            // If directory actually exists, search it for files
+            if (_fileSystem.Directory.Exists(folderComposite.FullPath))
             {
-                // Roslyn cannot compile xaml.cs files generated by xamarin. 
-                // Since the files are generated they should not be mutated anyway, so skip these files.
-                if (!file.EndsWith(".xaml.cs"))
+                foreach (var folder in _fileSystem.Directory.EnumerateDirectories(folderComposite.FullPath).Where(x => !_foldersToExclude.Contains(Path.GetFileName(x))))
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs]: Can we add the "virtual" paths here to the real paths? Just append the linked files dir if it also starts with the same string. We could reuse all other logic in this method for both real and linked files.

File: src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs [hunk 3]
```diff
@@ -96,15 +111,17 @@ private FolderComposite FindInputFiles(string path, List<string> filesToExclude,
...
+                {
+                    folderComposite.Add(FindInputFiles(folder, filesToExclude, compileIncludeLinkedFiles, folderComposite.RelativePath));
+                }
+
+                // Find all cs files on disk except .xaml.cs files as they do not need to be compiled
+                foreach (var file in _fileSystem.Directory.GetFiles(folderComposite.FullPath, "*.cs", SearchOption.TopDirectoryOnly).Where(f => !f.EndsWith(".xaml.cs")))
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs]: Same for getFiles, if we append the linked files to the files on the file system and return their real path they will be included correctly I guess.

**Comment by Mobrockers** [comment]
@richardwerkman I rewrote it in the way you suggested. Is this more clear now?

It doesn't work perfectly yet, the following structure is incorrectly identified:

```
<Compile Include="..\CustomAttribute - Copy (2).cs" Link="Trees\CustomAttribute.cs" />
<Compile Include="..\CustomAttribute - Copy (3).cs" Link="Bushes\CustomAttribute.cs" />
<Compile Include="..\CustomAttribute - Copy (5).cs" Link="Trees\Bushes\CustomAttribute.cs" />
```
..\CustomAttribute - Copy (5).cs get's included in all three folders.

Can you see the problem?

File: src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs [hunk 4]
```diff
@@ -46,23 +46,37 @@ public InputFileResolver(IFileSystem fileSystem, IProjectFileReader projectFileR
...
-            var projectUnderTestDir = Path.GetDirectoryName(result.ProjectUnderTestAnalyzerResult.ProjectFilePath);
-            foreach (var dir in ExtractProjectFolders(result.ProjectUnderTestAnalyzerResult, result.FullFramework))
+            var sharedProjectFilePaths = FindSharedProjects(projectAnalyzerResult.ProjectUnderTestAnalyzerResult);
+
+            var projectFolders = ExtractProjectFolders(projectAnalyzerResult.ProjectUnderTestAnalyzerResult, sharedProjectFilePaths);
+
+            // Search for compile linked source files in all linked shared projects
+            foreach (var sharedProjectFile in sharedProjectFilePaths)
+            {
+                compileIncludeLinkedFiles.ToList().AddRange(FindCompileLinkedFiles(sharedProjectFile));
```
Comment (jnyrup) [inline src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs]: `AddRange` is invoked on a newly created `List`, so `compileIncludeLinkedFiles` is never updated with the results from `FindCompileLinkedFiles`.

Comment (Mobrockers) [reply src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs]: Oh wow, thanks! Fixed.

**Comment by Mobrockers** [comment]
I missed this before but compile include linked supports wildcards and excludes. This needs to be redesigned.

**Pull request closed without merging**